src/node_platform.cc

```diff
@@ -452,23 +476,32 @@ void NodePlatform::CallDelayedOnWorkerThread(std::unique_ptr<Task> task,
 }
 
 
+IsolatePlatformDelegate* NodePlatform::ForIsolate(Isolate* isolate) {
+  Mutex::ScopedLock lock(per_isolate_mutex_);
+  auto data = per_isolate_[isolate];
+  CHECK(data.first);
+  return data.first;
+}
+
 std::shared_ptr<PerIsolatePlatformData>
-NodePlatform::ForIsolate(Isolate* isolate) {
+NodePlatform::ForNodeIsolate(Isolate* isolate) {
   Mutex::ScopedLock lock(per_isolate_mutex_);
-  std::shared_ptr<PerIsolatePlatformData> data = per_isolate_[isolate];
-  CHECK(data);
-  return data;
+  auto data = per_isolate_[isolate];
+  CHECK(data.second);
+  return data.second;
 }
 
 bool NodePlatform::FlushForegroundTasks(Isolate* isolate) {
-  return ForIsolate(isolate)->FlushForegroundTasksInternal();
+  return ForNodeIsolate(isolate)->FlushForegroundTasksInternal();
 }
 
-bool NodePlatform::IdleTasksEnabled(Isolate* isolate) { return false; }
+bool NodePlatform::IdleTasksEnabled(Isolate* isolate) {
+  return ForIsolate(isolate)->IdleTasksEnabled();
```

Comment:
Is this a potential performance regression? V8 calls this method in one or two possibly performance-sensitive places; the extra lookup might be noticeable.

Reply:
It's possible. Are there any systems in place to test? My intuition is that it's fine, everything should be very cpu pipeline-friendly.

Reply:
Chromium does something similar: https://github.com/chromium/chromium/blob/dae22e4bf16ea6bfecb1696f28d83462b842f26f/gin/v8_platform.cc#L443

I don't know if the lookups are comparable though. Chrome's `PerIsolateData::From` looks up the per-isolate data in an embedder field which is just one dereference vs node's `unordered_map::find`.

Reply:
I'm not sure if our benchmark suite has anything that would really stress-test that because I expect the performance issues to be around concurrent (re)compilation and idle scavenging but maybe `benchmark/worker/echo.js` with `workers=... n=...`, for high value of `workers` (1e2, 1e3, etc.) en low values of `n` (e.g. 1e3)?

Reply:
So my first question is "how many times is this method called"? Running echo.js with n=1000 workers=1000 takes ~185 seconds on my machine and results in ~14400 calls to `IdleTasksEnabled`, about ~78 per second. That's not nearly common enough for a method like this where I'd consider it performance critical, but I'll continue.

I compared a few variants of this branch against the merge base and came up with the following results:

As is in this PR: +0.65%
Cached result: +0.17%
Hard coded return false (in NodePlatform): -0.25%
Hard coded return false (in virtual IdleTasksEnabled): -0.27%

For the "cached result" variant I applied the following patch on top:
```diff
diff --git a/src/node.h b/src/node.h
index db24fca65e..caf9a90fe3 100644
--- a/src/node.h
+++ b/src/node.h
@@ -256,6 +256,11 @@ class NODE_EXTERN IsolatePlatformDelegate {
  public:
   virtual std::shared_ptr<v8::TaskRunner> GetForegroundTaskRunner() = 0;
   virtual bool IdleTasksEnabled() = 0;
+ private:
+  friend class NodePlatform;
+  bool idle_tasks_enabled_ = false;
+  void CacheIdleTasksEnabled() { idle_tasks_enabled_ = IdleTasksEnabled(); }
+  bool IdleTasksEnabledCached() { return idle_tasks_enabled_; }
 };
 
 class NODE_EXTERN MultiIsolatePlatform : public v8::Platform {
diff --git a/src/node_platform.cc b/src/node_platform.cc
index d875112d6f..8ed1425dd7 100644
--- a/src/node_platform.cc
+++ b/src/node_platform.cc
@@ -496,7 +496,7 @@ bool NodePlatform::FlushForegroundTasks(Isolate* isolate) {
 }
 
 bool NodePlatform::IdleTasksEnabled(Isolate* isolate) {
-  return ForIsolate(isolate)->IdleTasksEnabled();
+  return ForIsolate(isolate)->IdleTasksEnabledCached();
 }
 
 std::shared_ptr<v8::TaskRunner>
```

This approach would save one virtual function dispatch. Though the Chrome guys don't seem to care about an extra virtual dispatch considering the `IdleTasksEnabled` call in Chromium I linked above is virtual as well.

Overall the differences seem pretty negligible and within a reasonable margin of error. I think these differences fall short of "observable" and my instinct is that this won't affect performance.

I'm also happy to remove the `IdleTasksEnabled` feature entirely and leave it hardcoded as false because I personally don't need it. That might be throwing the baby out with the bathwater though.

Reply:
I did some benchmarking of my own and I see numbers on the same order of magnitude (between 15 and 100 calls/second) so yes, I agree it probably won't make an observable difference.

For the interested:
<details>
worker/echo.js n=1000 sendsPerBroadcast=1 payload="string" workers=1000: 18.03718620762461
1110 15.7/s
worker/echo.js n=1000 sendsPerBroadcast=10 payload="string" workers=1000: 6.691423850053621
3827 23.3/s
worker/echo.js n=1000 sendsPerBroadcast=1 payload="object" workers=1000: 23.283101204466668
1324 22.9/s
worker/echo.js n=1000 sendsPerBroadcast=10 payload="object" workers=1000: 13.985931667363088
8060 92.9/s
</details>